xs/src/libslic3r/SupportMaterial.cpp

```diff
@@ -0,0 +1,1225 @@
+    auto interface_layers_num = object_config->support_material_interface_layers.value;
+
+    for (int layer_id = 0; layer_id < support_z.size(); layer_id++) {
+        auto z = support_z[layer_id];
+
+        if (contact.count(z) <= 0)
+            continue;
+        Polygons &_contact = contact[z];
+
+        // Count contact layer as interface layer.
+        for (int i = layer_id - 1; i >= 0 && i > layer_id - interface_layers_num; i--) {
+            auto _z = support_z[i];
+            auto overlapping_layers = this->overlapping_layers(i, support_z);
+            vector<coordf_t> overlapping_z;
+            for (auto z_el : overlapping_layers)
+                overlapping_z.push_back(support_z[z_el]);
+
+            // Compute interface area on this layer as diff of upper contact area
+            // (or upper interface area) and layer slices.
+            // This diff is responsible of the contact between support material and
+            // the top surfaces of the object. We should probably offset the top
+            // surfaces vertically before performing the diff, but this needs
+            // investigation.
+            Polygons ps_1;
+            append_to(ps_1, _contact); // clipped projection of the current contact regions.
+            append_to(ps_1, interface[i]); // interface regions already applied to this layer.
+
+            Polygons ps_2;
+            for (auto el : overlapping_z) {
+                if (top.count(el) > 0)
+                    append_to(ps_2, top[el]); // top slices on this layer.
+                if (contact.count(el) > 0)
+                    append_to(ps_2, contact[el]); // contact regions on this layer.
+            }
+
+            _contact = interface[i] = diff(
+                ps_1,
+                ps_2,
+                true
+            );
+        }
+    }
+    return interface;
+}
+
+void
+SupportMaterial::generate_bottom_interface_layers(const vector<coordf_t> &support_z,
+                                                  map<int, Polygons> &base,
+                                                  map<coordf_t, Polygons> &top,
+                                                  map<int, Polygons> &interface)
+{
+    // If no interface layers are allowed, don't generate bottom interface layers.
+    if (object_config->support_material_interface_layers.value == 0)
+        return;
+
+    auto area_threshold = interface_flow.scaled_spacing() * interface_flow.scaled_spacing();
+
+    // Loop through object's top surfaces. TODO CHeck if the keys are sorted.
+    for (auto &top_el : top) {
+        // Keep a count of the interface layers we generated for this top surface.
+        int interface_layers = 0;
+
+        // Loop through support layers until we find the one(s) right above the top
+        // surface.
+        for (int layer_id = 0; layer_id < support_z.size(); layer_id++) {
+            auto z = support_z[layer_id];
+            if (!z > top_el.first) // next unless $z > $top_z;
```

Comment:
Compiler warning: the `!` is only being applied to `z`, so the expression looks like `if ((!z) > top_el.first)` and not `if (!(z > top_el.first))`.